llvm_passes/HipPromoteInts.cpp

```diff
@@ -0,0 +1,377 @@
+    Value *RHS = getPromotedValue(BinOp->getOperand(1));
+
+    Value *NewInst;
+    if (NeedsPromotion) {
+      // Create operation in promoted type
+      NewInst = Builder.CreateBinOp(BinOp->getOpcode(), LHS, RHS);
+    } else {
+      // For operations that should stay in original type
+      if (LHS->getType() != BinOp->getType())
+        LHS = Builder.CreateTrunc(LHS, BinOp->getType());
+      if (RHS->getType() != BinOp->getType())
+        RHS = Builder.CreateTrunc(RHS, BinOp->getType());
+      NewInst = Builder.CreateBinOp(BinOp->getOpcode(), LHS, RHS);
+    }
+
+    LLVM_DEBUG(dbgs() << Indent << "  " << *I << "    ============> "
+                      << *NewInst << "\n");
+    PromotedValues[I] = NewInst;
+    Replacements.push_back(Replacement(I, NewInst));
+  } else if (isa<CallInst>(I)) {
+    CallInst *OldCall = cast<CallInst>(I);
+    // Create a new call with the same operands, but use promoted values where
+    // available
+    SmallVector<Value *, 8> NewArgs;
+    for (unsigned i = 0; i < OldCall->arg_size(); ++i) {
+      Value *OldArg = OldCall->getArgOperand(i);
+      Value *NewArg =
+          PromotedValues.count(OldArg) ? PromotedValues[OldArg] : OldArg;
+
+      // if the function expects a non-standard type, abort for now.
+      // TODO: if this assert is hit, we need to handle this case in the future
+      // by promoting the function arguments as well.
+      if (OldArg->getType() != NewArg->getType())
+        assert(false &&
+               "HipPromoteIntsPass: Function expects non-standard type");
+
+      NewArgs.push_back(NewArg);
+    }
+
+    CallInst *NewCall = CallInst::Create(OldCall->getFunctionType(),
+                                         OldCall->getCalledOperand(), NewArgs,
+                                         OldCall->getName(), OldCall);
+    NewCall->setCallingConv(OldCall->getCallingConv());
+    NewCall->setAttributes(OldCall->getAttributes());
+
+    LLVM_DEBUG(dbgs() << Indent << "  " << *I << "    ============> "
+                      << *NewCall << "\n");
+    PromotedValues[I] = NewCall;
+    Replacements.push_back(Replacement(I, NewCall));
+  } else {
+    assert(false && "HipPromoteIntsPass: Unhandled instruction type");
+  }
+}
+
+static void promoteChain(Instruction *OldI, Type *NonStdType, Type *PromotedTy,
+                         SmallPtrSetImpl<Instruction *> &Visited,
+                         SmallVectorImpl<Replacement> &Replacements,
+                         SmallDenseMap<Value *, Value *> &PromotedValues,
+                         unsigned Depth = 0) {
+  // If we've already processed this instruction, just return
+  if (!Visited.insert(OldI).second) {
+    // If we have a promoted value for this instruction, use it
+    if (PromotedValues.count(OldI)) {
+      LLVM_DEBUG(dbgs() << std::string(Depth * 2, ' ')
+                        << "Already processed: " << *OldI << "\n");
+      return;
+    }
+    return;
```

Comment:
Could be simplified:
```suggestion
    if (PromotedValues.count(OldI))
      LLVM_DEBUG(dbgs() << std::string(Depth * 2, ' ')
                        << "Already processed: " << *OldI << "\n");
    return;
```